src/store/oauth/sagas.ts

```diff
@@ -0,0 +1,45 @@
+import { take, put } from 'redux-saga/effects'
+
+import {
+  REQUEST_NATIVE_OPEN_POPUP,
+  nativeOpenPopup,
+  SET_CREDENTIALS,
+  SetCredentialsAction,
+  RequestNativeOpenPopupAction
+} from './actions'
+import { Credentials } from './types'
+
+let resolve: (c: Credentials) => void
+let reject: (e: Error) => void
+
+function* watchRequestNativeOpenPopup() {
+  const action: RequestNativeOpenPopupAction = yield take(
+    REQUEST_NATIVE_OPEN_POPUP
+  )
+  resolve = action.resolve
+  reject = action.reject
+  yield put(nativeOpenPopup(action.url, action.provider))
+}
+
+function* watchSetCredentials() {
+  const action: SetCredentialsAction = yield take(SET_CREDENTIALS)
+
+  if (!action.credentials.error) {
+    if (resolve) {
+      resolve(action.credentials)
+    }
+  } else {
+    if (reject) {
```

Comment:
this flow is kind of confusing to me. i think i get it now, but maybe we should have some sort of invariant on resolve & reject being set. what if we put the watchSetCredentials inside the watchRequestNativeOpenPopup?

Also, should these be takeLatest or takeEvery? I think they'd only be allowed to run once the way they're written

Reply:
Ah of course, I forgot how sagas worked for a minute

Good callout, I think this is way cleaner. Is this what you had in mind? https://github.com/AudiusProject/audius-mobile-client/pull/198/commits/f03462eff81a4c1a6ef6220097f70a8a8fe993ea

Reply:
yep!